### ISSUE TITLE
Incorrect DNS response size

### ISSUE THREAD

**Issue by ltagliamonte-dd** [hashicorp/consul#9945]
## Context:
In our infrastructure we use [coredns](https://coredns.io/) in front of consul agent for DNS protocol.

Meanwhile investigating an issue we are facing with [coredns](https://github.com/coredns/coredns/issues/4547)
I found that the problem of the issue isn't in coredns but in Consul.

When EDNS is in use and buffsize is specified at 2048 bytes, Consul response size on the wire is 2058 bytes, and causes coredns to memory overflow.
The instructions, and tcpdump captures on how to replicate the bug, are all present in the first comment of this [issue](https://github.com/coredns/coredns/issues/4547)

## Consul code:
 Looking at the code (adding a few prints over the DNS request path) it appears that the problem is in how the [maxSize](https://github.com/hashicorp/consul/blob/566efad2e7eccce8b581150b61713ea065f0b43f/agent/dns.go#L1108) is calculated when EDNS is requested:
```
before truncation 4135
before extra 2039
after extra 2039
before setEdns 2039
before wire 2050
```

the truncation of the DNS response happens before [setEdns](https://github.com/hashicorp/consul/blob/566efad2e7eccce8b581150b61713ea065f0b43f/agent/dns.go#L500) and doesn't take  into account the 11 bytes that EDNS will add to the response before the wire.
As matter of fact if I modify [maxSize](https://github.com/hashicorp/consul/blob/566efad2e7eccce8b581150b61713ea065f0b43f/agent/dns.go#L1108) to be:
```
maxSize = int(size) - 11
```
The issue on coredns size is fixed.

**Comment by ltagliamonte-dd** [comment]
So far i found 2 possible solutions to fix this problem, but I am not really happy with any of those.
1. First one introduce a flag `EmptyOnTruncate` [code](https://github.com/hashicorp/consul/compare/master...ltagliamonte-dd:empty_on_truncate?expand=1)
2. Trim the response again before sending it down to the wire [code](https://github.com/hashicorp/consul/compare/master...ltagliamonte-dd:trim_at_the_end?expand=1)
##### Rationale behind the first option
Most DNS servers on truncate just send an empty response, with the Truncate flag.
Not happy with this one because we have the truncate info available after we have done already a lot of work to process the request, kinda feel like waste of processing.

##### Rationale behind the second option
The boolean flag [ecsGlobal](https://github.com/hashicorp/consul/blob/566efad2e7eccce8b581150b61713ea065f0b43f/agent/dns.go#L497) flag is strictly coupled with the processing of the request so we can't invoke setEDNS function before processing the request.
On the other end the dispatch function already calls the `trimDNSResponse` in a couple of cases, that so far i haven't found a way to de coupled as well.
I don't really like this solution as well because it doesn't really fixes the problem, and in the worse case (that didn’t happen during my test) the second trimDNSResponse could end-up in a second binary search to find the perfect size, that could impact the perfs on high latency sensitive applications.

Would love to get some feedbacks from you folks if I'm missing something and this can be fixed in any other/ better way.

**Comment by mootpt** [comment]
@dnephin does the team have any preference on @ltagliamonte-dd proposed [fixes](https://github.com/hashicorp/consul/issues/9945#issuecomment-809787158)? This has impacted our prod clusters a few times. Hoping to prioritize getting a fix in.

**Comment by dnephin** [comment]
I had a quick look. I noticed that `trimUDPResponse` already does this `edns.UDPSize` lookup. I think this suggests that the bug here is that `trimDNSResponse` is being called from the wrong place. I think an ideal solution would be to remove the existing calls to `trimDNSResponse` and replace them with new calls at the end of `handlePtr` and `handleQuery`, after `setEDNS`, but before the `WriteMsg`. This seems like the correct place to be truncating.

From what I can tell, these two functions (`handlePtr` and `handleQuery`) are the handlers for everything, so every request will call one of these two functions. It seems like those two functions have all the necessary data to be able to call `trimDNSResponse` and have it do the right thing.

I will say that I'm only mildly familiar with the dns server code, and not at all familiar with many of these DNS options, so I'm not sure how well that ideal solution will work. If there are problems or test failures that are caused by that change I can attempt to help sort through them.

**Comment by ltagliamonte-dd** [comment]
@dnephin I agree with you that `trimUDPResponse` should be invoked once at the very end before sending the DNS response out, but the code is a state that doesn't really allow this decoupling (without creating a big diff), that's why I opted for a suboptimal workaround: invoking for a second time `trimUDPResponse`.

Please let me know what do you think, this is a pressing issue for us.

**Comment by dnephin** [comment]
@ltagliamonte-dd I opened PR #10009, which does a bit of refactoring and I think may address this problem. There are a few things left to do, but it seems to be passing tests.

Do you have a test case to show that this change is working correctly for the requests that were previously not working? Would you be open to submitting a PR to add those tests? If you branch from `dnephin/trim-dns-response-with-edns` you should be able to open a PR against my branch, and I can get it included by merging the PR.

**Comment by ltagliamonte-dd** [comment]
Hello @dnephin thank you for looking into this, at this [link](https://github.com/coredns/coredns/issues/4547) there is the whole explanation on how to re-create the issue, the consul part is just this one:

```
cat <<'EOF' >> init-consul.sh
#!/bin/bash
for i in {0..254}
do
cat <<EOP > payload.json
{
  "Node": "foobar",
  "Address": "192.168.10.10",
  "Service": {
    "ID": "deploy-${i}",
    "Service": "deployment-service-web",
    "Address": "192.168.1.${i}",
    "Port": 8000
  }
}
EOP
curl --request PUT --data @payload.json http://127.0.0.1:8500/v1/catalog/register
done
rm payload.json
EOF
```

you can then make a DNS req using `dig @127.0.0.1 -p 8600 deployment-service-web.service.consul +edns +bufsize=2048 +ignore`
You should receive an answer with a total length <= 2048.
At the moment Consul is returning 2050bytes.

**Comment by ltagliamonte-dd** [comment]
dig output:
```
; <<>> DiG 9.10.6 <<>> @127.0.0.1 -p 8600 deployment-service-web.service.consul +edns +bufsize=2048 +ignore
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 4668
;; flags: qr aa tc rd ra; QUERY: 1, ANSWER: 124, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 2048 <--- ASK
;; QUESTION SECTION:
;deployment-service-web.service.consul. IN A
.....
.....
;; Query time: 1 msec
;; SERVER: 127.0.0.1#8600(127.0.0.1)
;; WHEN: Mon Apr 19 11:20:26 PDT 2021
;; MSG SIZE  rcvd: 2050 <--- RECEIVED
```

**Comment by ltagliamonte-dd** [comment]
Hello @dnephin, did you made any progress on this fix?

**Comment by ltagliamonte-dd** [comment]
Hello @dnephin any update on this issue?

**Comment by dnephin** [comment]
No updates yet unfortunately. I believe the TODO list in #10009 is still accurate. Still needs a bit more work before it is ready to merge.

**Comment by ltagliamonte** [comment]
@dnephin any updates on fixing this issues?
Anything i can try to help you with?

**Comment by ltagliamonte-dd** [comment]
@dnephin thank you fixing this!
Can we please cherry-pick to 1.8 and 1.9?

**Comment by dnephin** [comment]
Ya, I'll make sure that gets done before the next release

**Comment by dnephin** [comment]
Backports are #10641 and #10642

**Comment by ltagliamonte-dd** [comment]
@dnephin sorry for re-opening this issue but the backports haven't been added to any 1.8 release, i checked also latest 1.8.14 and the code isn't there, can you please add fix to 1.8.X ?

**Comment by blake** [comment]
Hi @ltagliamonte, #10642 backported the changes to the release/1.8.x branch. The specific commit is 6ccb653.

Consul 1.18.14 was released on 7/15/21/, but the backport PR was merged on 7/19/21. These changes will definitely be in the next 1.8.x release of Consul.

**Comment by ltagliamonte-dd** [comment]
thanks @blake any ETA?

**Comment by blake** [comment]
@ltagliamonte-dd, the tentative date for the next 1.8.x, 1.9.x, and 1.10.x patch releases is this Thursday (8/26/21). My apologies for not including this in my previous reply.

**Comment by ltagliamonte-dd** [comment]
Thank you @blake much appreciate

**Comment by hochuenw-dd** [comment]
Thanks! @blake, it seems somehow the image was not released. Can you take a look at the pipeline?

**Comment by blake** [comment]
@hochuenw-dd Sure, I'll look into why the official Docker image has not been updated. In the mean time, you can obtain the image from the other repo we manage, https://hub.docker.com/r/hashicorp/consul/tags?page=1&ordering=last_updated&name=1.8.

**Comment by hochuenw-dd** [comment]
Hi @blake, it seems 1.18.15 is still not released but 1.8 is updated.
